### PR TITLE
More features & breaking v2 changes

### LINKED ISSUE #16: Enhancement: Add polls to existing discussions
I have 3 times now forgot to create the poll on creation, and had to delete the old thread and create a new one.

I searched the flarum board and found several other users asking for this as well.

https://discuss.flarum.org/d/20586-friendsofflarum-polls/50
https://discuss.flarum.org/d/20586-friendsofflarum-polls/58
https://discuss.flarum.org/d/20586-friendsofflarum-polls/84
https://discuss.flarum.org/d/20586-friendsofflarum-polls/81

Thought I would aggregate and record interest.

### LINKED ISSUE #37: Drop reflar database compatibility
The migrations include special conditions for migrating the old reflar/polls table. This includes codes we must continue to maintain.

Those can be dropped. If anyone still requires them, they can just install an older version of fof/polls to perform the migration.

I suggest we remove that additional code, as well as the `FoF\Polls\Migrations\AbstractMigration` class and just use normal migrations. We can keep the same file names so nothing will break for anyone.

### LINKED ISSUE #38: Drop newPollVote websocket message
As of version 0.3.3, the `newPollVote` message on the Pusher websocket is no longer used. Instead a new `updatedPollOption` is used.

Version 0.3.3 continues to broadcast the old message for backward compatibility, but we should remove it in a future version.

This issue is there to collect any comment from someone who might be using that message as part of their own integration. If there's no comment, I think we can assume it can be safely removed without any harm.

### PR DESCRIPTION
**Pull request by dsevillamartin**
**🔴 THIS PR IS NOT BACKWARDS COMPATIBLE. DO NOT TEST IN PRODUCTION. 🔴**

_This PR is intended to be merged after #73. Will need to change base to v2 & rebased._

**Fixes #37**
**Fixes #38**
**Fixes #16**

**Changes proposed in this pull request:**
- Move voters button next to heading
- Fix no permission error after poll ends
- Create endpoint for new poll to add to existing post
    * Behavior & logic on post/discussion creation is the same - code is mostly shared
    * Prevent editing poll if user cannot edit post
    * Add policy to post for starting poll on said post - user must be able to edit post & start polls in discussion (tag scoped)
-  Drop 'newPollVote' websocket message
- Remove deprecated single vote API endpoint
- Drop ReFlar/polls database compatibility
- Rename 'canSeeVotes' to 'canSeeVoters' in `PollSerializer`
- Improve permissions
  - Added visibility scoping -- user must be able to see post
  - User must be able to edit post to edit poll

**Reviewers should focus on:**

**Screenshot**
image

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).

### CODE REVIEW

File: extend.php [hunk 1]
```diff
@@ -34,6 +34,7 @@
     new Extend\Locales(__DIR__.'/resources/locale'),
 
     (new Extend\Routes('api'))
+        ->post('/fof/polls', 'fof.polls.create', Controllers\CreatePollController::class)
```
Comment (dsevillamartin) [inline extend.php:37]: Should this maybe be `/posts/{id}/poll` instead ? It makes more sense for how the post ID is provided to the backend. 🤷

**Comment by Wlork** [comment]
Hello ! 

All the issues I reported on Flarum have been fixed. Congratulations ! 👍 

Unfortunately, I detected more bugs.... But I can see the light at the end of the tunnel ! 

These are bugs linked to the "end date" option. 

How can I explain it? 

On a discussion or a post, you create the first poll with an end date. It works 👍  

However, if you add another poll in the same post.... If you don't set an end date, it still creates a date automatically.... 
It's annoying, it shouldn't create a end date if you don't set anything.

It's a bug was hard for me to understand and reproduce, so I hope that I made clear :)

Also, It's not possible to remove the end date on any survey if you edit poll.

Thank

**Comment by dsevillamartin** [comment]
> However, if you add another poll in the same post.... If you don't set an end date, it still creates a date automatically....
It's annoying, it shouldn't create a end date if you don't set anything.

@Wlork Hm, I wasn't able to reproduce this... I created a poll with an end date, then on that same post, created another poll with no end date set. And it worked as intended - first poll has end date, second doesn't.

> Also, It's not possible to remove the end date on any survey if you edit poll.

This is probably from my changes in #72. I need to check if v1 allowed you to remove an end date - there were already checks so that you couldn't make it sooner than it was set so I guess I assumed it didn't.

**Comment by Wlork** [comment]
> Hm, I wasn't able to reproduce this... I created a poll with an end date, then on that same post, created another poll with no end date set. And it worked as intended - first poll has end date, second doesn't.

I forgot to mention that you need to activate the multiple choice option.

Sorry about that !

> This is probably from my changes in https://github.com/FriendsOfFlarum/polls/pull/72. I need to check if v1 allowed you to remove an end date - there were already checks so that you couldn't make it sooner than it was set so I guess I assumed it didn't.

I think it's good to be able to delete if needed :) Like the other options where we can enable/disable

**Comment by dsevillamartin** [comment]
> I forgot to mention that you need to activate the multiple choice option.

@Wlork I still can't reproduce this issue. Can you post a recording (gif or video upload works here on GitHub) showcasing the issue? 

> I think it's good to be able to delete if needed

I believe I have now reverted it back to the behavior in v1. Haven't built JS for it (too many commits with that 😅) but I'm not worried about that.

**Comment by Wlork** [comment]
> @Wlork I still can't reproduce this issue. Can you post a recording (gif or video upload works here on GitHub) showcasing the issue?

Hello @dsevillamartin :)

I don't have this bug anymore. Or maybe I just can't reproduce it :sweat_smile:

> I believe I have now reverted it back to the behavior in v1. Haven't built JS for it (too many commits with that sweat_smile) but I'm not worried about that.

I'm not sure if I've understood you correctly, but I still can't cancel existing poll's end date :(

After the last update, I was able to run several tests. 

I detected another bug.

As a member : 

1- Create a discussion or a post with a poll
2- Reply to the discussion (or post), with or without a poll, it doesn't matter
3- Press F5
4- The edit and delete buttons for the previous poll have disappeared.
5- and so on



Apart from these two bugs, everything seems to be running smoothly on my side.

Once you've corrected these issues, I'll redo a complete test from A to Z :)

**Comment by dsevillamartin** [comment]
> I'm not sure if I've understood you correctly, but I still can't cancel existing poll's end date :(

I didn't build the JS assets for it, so unless you do it in your vendor folder, it won't be fixed for you. I'm not worried about it still being a bug though.

>  2- Reply to the discussion (or post), with or without a poll, it doesn't matter

I've made it so that you need to be able to edit a post to edit and add polls. I feel like this makes sense, as polls are a part of the post. So this is intended behavior on my part. You have it set so you can no longer edit a post once it's been replied to, and that should (in my opinion) extend to its polls.

What I'm wondering now is whether I took the "edit own polls" permission into consideration during this. I may add add a new permission for editing polls even if the user can no longer edit the post itself.

**Comment by Wlork** [comment]
> I didn't build the JS assets for it, so unless you do it in your vendor folder, it won't be fixed for you. I'm not worried about it still being a bug though.

OK, I understand :) 

> I've made it so that you need to be able to edit a post to edit and add polls. I feel like this makes sense, as polls are a part of the post. So this is intended behavior on my part. You have it set so you can no longer edit a post once it's been replied to, and that should (in my opinion) extend to its polls.

The butttons being gone is good - them not being gone beforehand is something that can't really be helped. The Flarum post "edit" button remains until refresh as well, even though the user no longer has the permission to edit that post.

What I'm wondering now is whether I took the "edit own polls" permission into consideration during this. I may add add a new permission for editing polls even if the user can no longer edit the post itself.

I just checked and I didn't give permission to edit their posts in my forum test.... That's why I had this "issue". I changed it and it's fixed :+1: 

I'm sorry :(

Good.... We'll have to see how it works out in practice, but as far as I'm concerned, everything's OK :) 
Would you like me to test anything in particular ? :)

**Comment by dsevillamartin** [comment]
@Wlork I've split permissions for poll editing for more control (polls you create vs polls others create on your posts). And built the JS.

There's still some bugs and features that would be helpful, but I think this is good for an initial v2 release. If you're planning on re-testing everything, let me know & I'll wait. Otherwise I think I'm ready to merge & release #73 and this.

Thank you for all your testing again.

**Comment by Wlork** [comment]
Hello @dsevillamartin 

I did all the tests from scratch :)

I can delete the end date when I edit a poll. The bug has disappeared 👍 The bug was probably due to the numerous updates....

Apart from that, I haven't found any major bugs :) 

Just : 

https://github.com/FriendsOfFlarum/polls/pull/74#issuecomment-1629516255

I saw this bug once. It appears when you add a poll on an existing post with at least one poll. But it's very random. I don't know when exactly this bug appears.... We'll have to see in practice with several users, I hope !

> Thank you for all your testing again.

Thank you for taking the time to develop these features !

I'll say it again, it's a really good extension you've developed. If you need any further tests, don't hesitate to let me know 👍

**Pull request merged**